pulsar/transaction_impl.go

```diff
@@ -0,0 +1,235 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package pulsar
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	pb "github.com/apache/pulsar-client-go/pulsar/internal/pulsar_proto"
+	"github.com/apache/pulsar-client-go/pulsar/log"
+	uAtomic "go.uber.org/atomic"
+)
+
+type TxnState int32
+type void struct{}
+type subscription struct {
+	topic        string
+	subscription string
+}
+
+type transaction struct {
+	sync.Mutex
+	txnID                    TxnID
+	state                    TxnState
+	tcClient                 *transactionCoordinatorClient
+	registerPartitions       map[string]void
+	registerAckSubscriptions map[subscription]void
+	/**
+	* opsFlow Wait all the operations of sending and acking messages with the transaction complete
+	* by reading msg from the chan.
+	* opsCount is record the number of the uncompleted operations.
+	* opsFlow
+	*        Write:
+	* 		      1. When the transaction is created, a new empty struct{} will be written to opsFlow chan.
+	*			  2. When the opsCount decrement from 1 to 0, a new empty struct{} will be written to opsFlow chan.
+	* 			  3. When get a retryable error after committing or aborting the transaction,
+	*	             a new empty struct{} will be written to opsFlow chan.
+	*        Read:
+	*			  1. When the transaction is committed or aborted, an empty struct{} will be read from opsFlow chan.
+	*			  2. When the opsCount increment from 0 to 1, an empty struct{} will be read from opsFlow chan.
+	 */
+	opsFlow   chan void
```

Comment:
I agree with @shibd. The `opsFlow` is hard to understand. We usually create a channel to handle all events in a separated goroutine, and send events to this channel from various goroutines.

See https://go.dev/blog/codelab-share

Reply:
The `opsFlow` has two effects:
1. It can be a protection mechanism to prevent sending or confirming messages with an ended transaction.
     Using the end transaction to send a message makes it easy to cause the stable position to not move for a long time. And this causes a read delay.
2. It can be used as a future to wait for all operations complete before committing or aborting transactions.

Reply:
oh, I see.

When `registerProducerTopic` and `registerAckTopic` operations are being performed, commits and aborts are also not allowed, right?

I see that the future of these Java operations is linked:
- `registerAckedTopic` future -> `registerAckOp`
- `registerProducedTopic` future -> `registerSendOp`


I'm not quite sure what to do with the go client next, maybe you can add some more complete tests?

`registerSendOrAckOp` and `endSendOrAckOp` and `registerProducerTopic` and `registerAckTopic`  these complete a test scenario overlay.

Reply:
There are some tests for `registerSendOrAckOp` and `endSendOrAckOp` and `registerProducerTopic` and `registerAckTopic`.
* `TestRegisterOpAndEndOp`
* `TestRegisterTopic`

Reply:
Overall LGTM, I'll wait to see how subsequent PRs use these methods together.